### DISCUSSION TITLE
Colors: Support unique color by value across fields

### DISCUSSION THREAD

**Discussion by colega** [grafana/grafana#38306] [Dashboards, Panels & Visualizations]
**What would you like to be added**: a "Classic palette" which has colors assigned by value, like in [other color schemes](https://user-images.githubusercontent.com/1511481/122519491-cfd89b00-d012-11eb-93ae-3774fc38c4c0.png). Or, even better, assigning \[pseudo-\]unique values to each new value in the series [instead of value percentage-based](http://github.com/grafana/grafana/blob/7f882eea05d7054407cf2c1fa05cbfe38414aa05/packages/grafana-data/src/field/fieldColor.ts#L195-L195).

**Why is this needed**: in the new _State timeline_ panel introduced in Grafana 8, choosing the palettes that assign colors by value makes more sense than selecting different colors for different series, since it allows us to visually differentiate the values. 

Consider the difference between this panel representing a version history: 

![image](https://user-images.githubusercontent.com/1511481/122554710-a1ba8180-d039-11eb-9090-d1fd0389a889.png)

And this one:

![image](https://user-images.githubusercontent.com/1511481/122554861-ca427b80-d039-11eb-9b98-23e3e1cf138a.png)

The latter one is not perfect, but at least it's _visually more distinguishable_. It represents some issues however:
- The value in the middle gets assigned the last color, so we can't really see that there are three values there (three text labels are the only clue, I actually [opened a bug report for this first](https://github.com/grafana/grafana/issues/35934), since I thought it was the same color, but it's just very similar).
- Regardless the _color scheme "by value"_ we choose, it adds a semantic to the value (in this case value is a timestamp, but newer doesn't mean it's better or worse, it's just a different value)

As a proof of concept, I tried this (this is my first contact with TypeScript :D)

```diff
diff --git packages/grafana-data/src/field/fieldColor.ts packages/grafana-data/src/field/fieldColor.ts
index ffbdf95ac2..1c90287850 100644
--- packages/grafana-data/src/field/fieldColor.ts
+++ packages/grafana-data/src/field/fieldColor.ts
@@ -48,6 +48,15 @@ export const fieldColorModeRegistry = new Registry<FieldColorMode>(() => {
         return theme.visualization.palette;
       },
     }),
+    new FieldColorSchemeMode({
+      id: 'unique',
+      name: 'Unique',
+      isContinuous: false,
+      isByValue: true,
+      getColors: (theme: GrafanaTheme2) => {
+        return theme.visualization.palette;
+      },
+    }),
     new FieldColorSchemeMode({
       id: 'continuous-GrYlRd',
       name: 'Green-Yellow-Red',
@@ -176,6 +185,18 @@ export class FieldColorSchemeMode implements FieldColorMode {
   getCalculator(field: Field, theme: GrafanaTheme2) {
     const colors = this.getColors(theme);

+    if (this.id === 'unique') {
+      let assigned: Map<number, number> = new Map<number, number>();
+      let next = 0;
+      return (value: number, _percent?: number, _threshold?: Threshold) => {
+        if (!assigned.has(value)) {
+          assigned.set(value, next);
+          next = (next + 1) % colors.length;
+        }
+        return colors[assigned.get(value) || 0];
+      };
+    }
+
     if (this.isByValue) {
       if (this.isContinuous) {
         return (_: number, percent: number, _threshold?: Threshold) => {
```

And it rendered me pretty much what I was looking for:

![image](https://user-images.githubusercontent.com/1511481/122557778-83568500-d03d-11eb-8f62-c64530703083.png)

And with more values:

![image](https://user-images.githubusercontent.com/1511481/122557845-9b2e0900-d03d-11eb-8a0a-7ccccfe973ac.png)

(This looks like a [Volkwagen Polo Harlequin](https://www.google.com/search?q=vw+polo+harlequin))

--- 

So, do we want this to be added? And if yes, am I on the right path? Which side-effects have I missed (I see lots of methods related to color mode, and @torkelo mentioned possible issues when streaming data)

**Comment by torkelo** [comment]
similar to https://github.com/grafana/grafana/issues/35945

**Comment by Legion2** [comment]
@colega this is exactly what I need as described in #35945

**Comment by colega** [comment]
Yes, I didn't add the requirement to assign same color to same value in different series, however but that would be definitely an improvement too.

@torkelo do we want to merge both issues into #35945 and close this one?

**Comment by torkelo** [comment]
we can keep this one and close the other, but we can add to this one that it would be good to support unique colors across fields as well (a bit more tricky)

**Comment by pkolyvas** [comment]
We're in the process of moving Grafana feature requests to [GitHub Discussions](https://community.grafana.com/t/were-moving-grafana-feature-requests-to-github-discussions/50341/3). This issue is being moved to [Discussions](https://github.com/grafana/grafana/discussions) or has been moved from our issues list, depending on where you're seeing this message. 

 We hope to use the discussion thread to continue building use cases, and additional evidence based on engagement/voting, in order to help us prioritize this feature request later on down the road. 

As with all things Grafana, we value your input and participation in this project and appreciate your patience during this transitional period. Looking forward to the discussion!

**Comment by Legion2** [comment]
The State Timeline is my new favorite visualisation tool. Choosing good coloring is very important to make information quickly understandable by humans. Here are two good examples for State Timeline visualisations of numerical and boolean values:
![image](https://user-images.githubusercontent.com/12752323/130921957-4dba2c40-c529-48e3-9c47-244ebc4cff12.png)

![image](https://user-images.githubusercontent.com/12752323/130921871-fb37458c-b7c7-4eca-adfa-e5f2bbb8d21e.png)

However for string based values currently there is no good coloring option in grafana:
Classic palette:
![image](https://user-images.githubusercontent.com/12752323/130922574-e16a10ed-0e0a-4994-8f5f-e7840be24b8f.png)
Green-Yellow-Red:
![image](https://user-images.githubusercontent.com/12752323/130922848-815e53aa-0322-439e-89e5-921ddfb3612e.png)

As a workaround for enum like string values (fixed small set of strings) a value mapping can be defined, to assign each string a color. This only works for a small set of strings and if the strings are known beforehand.

It would be better if grafana can automatically assign each string from the query result a unique color and use it across all series. The mapping from string to color should then also be shown in the legend (currently value mappings are not shown in the legend).

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). We are migrating this discussion to an issue and closing the discussion. The issue is #82979. Feel free to continue the discussion around this there. Thank you!
